modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernel.C

```diff
@@ -0,0 +1,222 @@
+
+  return params;
+}
+
+CZMInterfaceKernel::CZMInterfaceKernel(const InputParameters & parameters)
+  : InterfaceKernel(parameters),
+    _disp_index(getParam<unsigned int>("disp_index")),
+    _disp_1(_mesh.dimension() >= 2 ? coupledValue("disp_1") : _zero),
+    _disp_1_neighbor(_mesh.dimension() >= 2 ? coupledNeighborValue("disp_1_neighbor") : _zero),
+    _disp_2(_mesh.dimension() >= 3 ? coupledValue("disp_2") : _zero),
+    _disp_2_neighbor(_mesh.dimension() >= 3 ? coupledNeighborValue("disp_2_neighbor") : _zero),
+    _disp_1_var(coupled("disp_1")),
+    _disp_1_neighbor_var(coupled("disp_1_neighbor")),
+    _disp_2_var(coupled("disp_2")),
+    _disp_2_neighbor_var(coupled("disp_2_neighbor")),
+    // residual and jacobian coefficients are material properties and represents
+    // the residual and jacobain of the traction sepration law wrt the displacement jump.
+    _residual(getParam<std::string>("residual")),
+    _jacobian(getParam<std::string>("jacobian")),
+    _throw_exception("throw_exception_mp"),
+    _ResidualMP(getMaterialProperty<RealVectorValue>(_residual)),
+    _JacobianMP(getMaterialProperty<RankTwoTensor>(_jacobian))
+{
+  if (!parameters.isParamValid("boundary"))
+  {
+    mooseError("In order to use  CZMInterfaceKernel ,"
+               " you must specify a boundary where it will live.");
+  }
+}
+
+Real
+CZMInterfaceKernel::computeQpResidual(Moose::DGResidualType type)
+{
+
+  Real r = _ResidualMP[_qp](_disp_index);
+
+  switch (type)
+  {
+    // [test_slave-test_master]*T where T repsents the traction.
+    // the + and - signs below are in accordance with this convention
+    case Moose::Element:
+      r *= -_test[_i][_qp];
+      break;
+
+    case Moose::Neighbor:
+      r *= _test_neighbor[_i][_qp];
+      break;
+  }
+
+  return r;
+}
+
+Real
+CZMInterfaceKernel::computeQpJacobian(Moose::DGJacobianType type)
+{
+  // retrieve the diagonal jacobain coefficient dependning on the disaplcement
+  // component (_disp_index) this kernel is working on
+  Real jac = _JacobianMP[_qp](_disp_index, _disp_index);
+
+  switch (type)
+  {
+    // (1) and (-1) terms in parenthesis are the derivatives of \deltaU with respect to slave and
+    // master variables to make the code easier to understand the trailing + and - signs are
+    // inherited directly from the reidual equation
+    case Moose::ElementElement:
+      jac *= -_test[_i][_qp] * _phi[_j][_qp] * (-1);
+      break;
+
+    case Moose::ElementNeighbor:
+      jac *= -_test[_i][_qp] * _phi_neighbor[_j][_qp] * (1);
+      break;
+
+    case Moose::NeighborElement:
+      jac *= _test_neighbor[_i][_qp] * _phi[_j][_qp] * (-1);
+      break;
+
+    case Moose::NeighborNeighbor:
+      jac *= _test_neighbor[_i][_qp] * _phi_neighbor[_j][_qp] * (1);
+      break;
+  }
+
+  return jac;
+}
+
+Real
+CZMInterfaceKernel::computeQpOffDiagJacobian(Moose::DGJacobianType type, unsigned int jvar)
```

Comment:
look at `StressDivergenceTensors::computeQpOffDiagJacobian(unsigned int jvar)`

Reply:
I've cleaned this  a lot.